fundamentals/middleware/rate-limit/WebRateLimitAuth/Program.cs

```diff
@@ -0,0 +1,454 @@
+                     autoReplenishment: myOptions.autoReplenishment)));
+
+app.MapGet("/", () => Results.Ok($"Token Limiter {GetTicks()}"))
+                           .RequireRateLimiting(tokenPolicy);
+
+app.Run();
+// </snippet_token>
+#elif FIRST
+// <snippet_1>
+using System.Globalization;
+using System.Net;
+using System.Threading.RateLimiting;
+using Microsoft.AspNetCore.Identity;
+using Microsoft.AspNetCore.RateLimiting;
+using Microsoft.EntityFrameworkCore;
+using Microsoft.Extensions.Logging.Abstractions;
+using Microsoft.Extensions.Options;
+using WebRateLimitAuth;
+using WebRateLimitAuth.Data;
+using WebRateLimitAuth.Models;
+
+var builder = WebApplication.CreateBuilder(args);
+
+var connectionString = builder.Configuration.GetConnectionString("DefaultConnection") ??
+    throw new InvalidOperationException("Connection string 'DefaultConnection' not found.");
+builder.Services.AddDbContext<ApplicationDbContext>(options =>
+    options.UseSqlServer(connectionString));
+builder.Services.AddDatabaseDeveloperPageExceptionFilter();
+
+builder.Services.AddDefaultIdentity<IdentityUser>(options => options.SignIn.RequireConfirmedAccount = true)
+    .AddEntityFrameworkStores<ApplicationDbContext>();
+
+builder.Services.Configure<MyRateLimitOptions>(
+    builder.Configuration.GetSection(MyRateLimitOptions.MyRateLimit));
+
+builder.Services.AddRazorPages();
+builder.Services.AddControllersWithViews();
+
+var app = builder.Build();
+
+if (app.Environment.IsDevelopment())
+{
+    app.UseMigrationsEndPoint();
+}
+else
+{
+    app.UseExceptionHandler("/Error");
+    app.UseHsts();
+}
+
+app.UseHttpsRedirection();
+app.UseStaticFiles();
+
+app.UseRouting();
+
+// <snippet>
+app.UseAuthentication();
+app.UseAuthorization();
+
+var userPolicyName = "user";
+var completePolicyName = "complete";
+var helloPolicy = "hello";
+var myOptions = new MyRateLimitOptions();
+var myConfigSection = app.Configuration.GetSection(MyRateLimitOptions.MyRateLimit);
+myConfigSection.Bind(myOptions);
+
+
+var options = new RateLimiterOptions()
+{
+    OnRejected = (context, cancellationToken) =>
+    {
+        if (context.Lease.TryGetMetadata(MetadataName.RetryAfter, out var retryAfter))
+        {
+            context.HttpContext.Response.Headers.RetryAfter =
+            ((int)retryAfter.TotalSeconds).ToString(NumberFormatInfo.InvariantInfo);
+        }
+
+        context.HttpContext.Response.StatusCode = StatusCodes.Status429TooManyRequests;
+        context?.HttpContext?.RequestServices?.GetService<ILoggerFactory>()?
+                      .CreateLogger("Microsoft.AspNetCore.RateLimitingMiddleware")
+                      .LogWarning($"OnRejected: {GetUserEndPoint(context.HttpContext)}");
+
+        return new ValueTask();
+    }
+}
+    //.AddPolicy<string>(completePolicyName, 
+    //           new SampleRateLimiterPolicy(NullLogger<SampleRateLimiterPolicy>.Instance,
+    //           (IOptions < MyRateLimitOptions >)myConfigSection))
```

Comment:
@Kahbazi how do I pass `IOptions<MyRateLimitOptions>` here?

Reply:
@Kahbazi  Is there any advantage to showing `new SampleRateLimiterPolicy()`? Maybe we should just use ` .AddPolicy<string, SampleRateLimiterPolicy>(helloPolicy)` like we do on the next line.

Reply:
> how do I pass IOptions<MyRateLimitOptions> here?

When you have the `MyRateLimitOptions` instance, you can create it with `Microsoft.Extensions.Options.Options.Create<MyRateLimitOptions>(options)`.

> @Kahbazi Is there any advantage to showing new SampleRateLimiterPolicy()? Maybe we should just use .AddPolicy<string, SampleRateLimiterPolicy>(helloPolicy) like we do on the next line.

Yeah, I agree. Having just `.AddPolicy<string, SampleRateLimiterPolicy>(helloPolicy)` for sample is enough I think.